zebra-chain/src/keys/sapling.rs

```diff
@@ -7,17 +7,188 @@
+            .finish()
+    }
+}
+
+impl From<SpendingKey> for ProofAuthorizingKey {
+    /// For this invocation of Blake2b-512 as PRF^expand, t=1.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#saplingkeycomponents
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+    fn from(spending_key: SpendingKey) -> ProofAuthorizingKey {
+        let mut block = [0u8; 33];
+        block[33] = 1; // Last byte is t=1;
+
+        block[0..32].copy_from_slice(&spending_key.0[..]);
+
+        let mut hasher = Blake2b::new();
+        // TODO: check that this counts as personalization.
+        hasher.input("Zcash_ExpandSeed");
+        hasher.input(block);
+
+        Self::from(hasher.result())
+    }
+}
+
+/// Derived from a _SpendingKey_.
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct OutgoingViewingKey(pub [u8; 32]);
+
+impl fmt::Debug for OutgoingViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("OutgoingViewingKey")
+            .field(&hex::encode(&self.0))
+            .finish()
+    }
+}
+
+impl From<SpendingKey> for OutgoingViewingKey {
+    /// For this invocation of Blake2b-512 as PRF^expand, t=1.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#saplingkeycomponents
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+    fn from(spending_key: SpendingKey) -> OutgoingViewingKey {
+        let mut block = [0u8; 33];
+        block[33] = 2u8; // Last byte is t=2;
+
+        block[0..32].copy_from_slice(&spending_key.0[..]);
+
+        let mut hasher = Blake2b::new();
+        // TODO: check that this counts as personalization.
+        hasher.input("Zcash_ExpandSeed");
+        hasher.input(block);
+
+        Self(hasher.result()[0..31])
+    }
+}
+
+///
+pub type AuthorizingKey = redjubjub::PublicKeyBytes;
+
+impl fmt::Debug for AuthorizingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("AuthorizingKey")
+            .field(&hex::encode(&self.0))
+            .finish()
+    }
+}
+
+///
+pub type NullifierDerivingKey = redjubjub::PublicKeyBytes;
+
+///
+pub type IncomingViewingKey = redjubjub::Scalar;
+
+///
+#[derive(Copy, Clone, Display, Eq, PartialEq)]
+pub struct Diversifier(pub [u8; 11]);
+
+impl fmt::Debug for Diversifier {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("Diversifier")
+            .field(&hex::encode(&self.0))
+            .finish()
+    }
+}
+
+/// Derived from a _IncomingViewingKey_ and a _Diversifier_
+///
+/// Derived by multiplying the diversifier (converted to an affine
+/// point on the Jubjub curve) by the incoming view key:
+/// pkd = gd * ivk
+pub type TransmissionKey = redjubjub::PublicKeyBytes;
+
+/// Full Viewing Keys
+///
+/// For incoming viewing keys on the production network, the
+/// Human-Readable Part is “zviews”. For incoming viewing keys on the
+/// test network, the Human-Readable Part is “zviewtestsapling”.
```

Comment:
Add a reference to https://zips.z.cash/protocol/protocol.pdf#saplingfullviewingkeyencoding

Reply:
Note that for zcashd we only intend to expose ZIP 32 extended full viewing keys to users.

Reply:
For Sapling only, right?